In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# Suppress Warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

In [5]:
df = pd.read_csv(r"E:\Programming\Machine Learning\ML Algorithams\Regression\Ridge_Regression\Melbourne_housing_FULL.csv")
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [7]:
df.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [8]:
df.shape

(34857, 21)

In [10]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
df = df[cols_to_use]

In [12]:
df.shape

(34857, 15)

In [23]:
df.head()

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   
5      3         4019.0       2.5       3.0       2.0  1.0      94.0   
6      4         4019.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0              False  ...   
2       79.0000  1035000.0              False  ...   
4      150.0000  1465000.0              False  ...   
5      160.2564   850000.0              False  ...   
6      142.0000  1600000.0              False  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                False                              False   
2                                False                              False   
4                                False                              False   
5                                False                              False   
6                                False                              False   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                False                                  False   
2                                False                                  False   
4                                False                                  False   
5                                False                                  False   
6                                False                                  False   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                 False                                False   
2                                 False                                False   
4                                 False                                False   
5                                 False                                False   
6                                 False                                False   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                False                             False   
2                                False                             False   
4                                False                             False   
5                                False                             False   
6                                False                             False   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                            True                                   False  
2                            True                                   False  
4                            True                                   False  
5                            True                                   False  
6                            True                                   False  

[5 rows x 745 columns]

In [13]:
df.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [15]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
df[cols_to_fill_zero] = df[cols_to_fill_zero].fillna(0)

In [17]:
# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
df['Landsize'] = df['Landsize'].fillna(df.Landsize.mean())
df['BuildingArea'] = df['BuildingArea'].fillna(df.BuildingArea.mean())

In [18]:
df.isna().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Regionname          3
Propertycount       0
Distance            0
CouncilArea         3
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
Price            7610
dtype: int64

In [19]:
df.dropna(inplace=True)

In [20]:
df.isna().sum()

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Regionname       0
Propertycount    0
Distance         0
CouncilArea      0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
Price            0
dtype: int64

In [21]:
df.shape

(27244, 15)

Let's one hot encode the categorical features

In [22]:
df = pd.get_dummies(df, drop_first=True)
df.head()

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   
5      3         4019.0       2.5       3.0       2.0  1.0      94.0   
6      4         4019.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0              False  ...   
2       79.0000  1035000.0              False  ...   
4      150.0000  1465000.0              False  ...   
5      160.2564   850000.0              False  ...   
6      142.0000  1600000.0              False  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                False                              False   
2                                False                              False   
4                                False                              False   
5                                False                              False   
6                                False                              False   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                False                                  False   
2                                False                                  False   
4                                False                                  False   
5                                False                                  False   
6                                False                                  False   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                 False                                False   
2                                 False                                False   
4                                 False                                False   
5                                 False                                False   
6                                 False                                False   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                False                             False   
2                                False                             False   
4                                False                             False   
5                                False                             False   
6                                False                             False   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                            True                                   False  
2                            True                                   False  
4                            True                                   False  
5                            True                                   False  
6                            True                                   False  

[5 rows x 745 columns]

In [27]:
X = df.drop('Price', axis=1)
y = df['Price']

In [29]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=2)


In [30]:
from sklearn.linear_model import  LinearRegression
reg = LinearRegression().fit(X_train, y_train)

In [31]:
reg.score(X_test, y_test)

0.13853683161455688

In [32]:
reg.score(X_train, y_train)

0.6827792395792723

Here training score is 68% but test score is 13.85% which is very low

Normal Regression is clearly overfitting the data, let's try other models

Using Lasso (L1 Regularized) Regression Model

In [34]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(X_train, y_train)

Lasso(alpha=50, max_iter=100, tol=0.1)

In [35]:
lasso_reg.score(X_train, y_train)

0.6766985624766824

In [36]:
lasso_reg.score(X_test, y_test)

0.6636111369404488

Using Ridge (L2 Regularized) Regression Model

In [38]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(X_train, y_train)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [39]:
ridge_reg.score(X_train, y_train)

0.6622376739684328

In [40]:
ridge_reg.score(X_test, y_test)

0.6670848945194956

We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too